packages/docker-engine/0001-Add-ability-to-read-default-OCI-Spec-from-a-file.patch

```diff
@@ -0,0 +1,212 @@
++		defCaps := caps.DefaultCapabilities()
++		defCaps = append(defCaps, s.Process.Capabilities.Bounding...)
++		defCaps = getUniqueCapabilities(defCaps)
++
+ 		capabilities, err := caps.TweakCapabilities(
+-			caps.DefaultCapabilities(),
++			defCaps,
+ 			c.HostConfig.CapAdd,
+ 			c.HostConfig.CapDrop,
+ 			c.HostConfig.Privileged,
+@@ -171,6 +182,19 @@ func WithCapabilities(c *container.Container) coci.SpecOpts {
+ 	}
+ }
+ 
++func getUniqueCapabilities(defCaps []string) []string {
++	exists := make(map[string]bool, len(defCaps))
++	var uniqueCaps []string
++	for c := range defCaps {
++		if !exists[defCaps[c]] {
++			exists[defCaps[c]] = true
++			uniqueCaps = append(uniqueCaps, defCaps[c])
++		}
++	}
++
++	return uniqueCaps
++}
++
+ func resourcePath(c *container.Container, getPath func() (string, error)) (string, error) {
+ 	p, err := getPath()
+ 	if err != nil {
+@@ -1009,11 +1033,46 @@ func WithUser(c *container.Container) coci.SpecOpts {
+ 	}
+ }
+ 
++func loadDefaultOCISpec(filename string) (specs.Spec, error) {
++	var spec specs.Spec
++
++	baseSpec, errReadfile := os.ReadFile(filename)
++	// Fail when unable to read the base spec file
++	if errReadfile != nil {
++		return spec, fmt.Errorf("unable to read the base spec file provided in Daemon: %v", errReadfile)
++	}
++
++	errUnmarshall := json.Unmarshal(baseSpec, &spec)
++	if errUnmarshall != nil {
++		return spec, fmt.Errorf("unable to parse the base spec file provided in Daemon: : %s: %w", filename, errUnmarshall)
++	}
++
++	return spec, nil
++}
++
+ func (daemon *Daemon) createSpec(c *container.Container) (retSpec *specs.Spec, err error) {
+ 	var (
+ 		opts []coci.SpecOpts
+ 		s    = oci.DefaultSpec()
+ 	)
++
++	var filename string
++	runtime := daemon.configStore.GetRuntime(c.HostConfig.Runtime)
++	if runtime.BaseRuntimeSpec != "" {
++		// Change the default spec as per base-spec-file for runtime
++		filename = runtime.BaseRuntimeSpec
++	} else if daemon.configStore.BaseRuntimeSpec != "" {
++		// Change the default spec as per base-spec-file in Daemon.json
++		filename = daemon.configStore.BaseRuntimeSpec
++	}
++	if filename != "" {
++		spec, err := loadDefaultOCISpec(filename)
++		if err != nil {
++			return nil, err
++		}
++		s = spec
++	 }
++
+ 	opts = append(opts,
+ 		WithCommonOptions(daemon, c),
+ 		WithCgroups(daemon, c),
+@@ -1075,3 +1134,21 @@ func (daemon *Daemon) mergeUlimits(c *containertypes.HostConfig) {
+ 	}
+ 	c.Ulimits = ulimits
+ }
++
++// mergeBaseSpecUlimits merge the Ulimits from Base Spec file and update daemon defaults
++func (daemon *Daemon) mergeBaseSpecUlimits(s *coci.Spec) {
++	ulIdx := daemon.configStore.Ulimits
++
++	for _, ul := range s.Process.Rlimits {
++		rlimit := strings.ToLower(strings.TrimPrefix(ul.Type, "RLIMIT_"))
++		if _, exists := ulIdx[rlimit]; !exists {
++			ulIdx[rlimit] = &units.Ulimit{
++				Name: rlimit,
++				Hard: int64(ul.Hard),
++				Soft: int64(ul.Soft),
++			}
++		}
++
++	}
++	daemon.configStore.Ulimits = ulIdx
++}
```

Comment:
I'd prefer to do this work in `mergeUlimits` - it's hard for me to follow what's meant to happen here, with two different function calls that both treat `daemon.configStore.Ulimits` as a source of truth.

Comments would be very helpful. I'm not really sure that the logic is correct. Unit tests would help too; I doubt this could go upstream without them because the three-way merge of defaults for both rlimits and capabilities is complicated.